include/oneapi/dpl/internal/reduce_by_segment_impl.h

```diff
@@ -138,6 +138,288 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
 }
 
 #if _ONEDPL_BACKEND_SYCL
+
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Range4,
+          typename _BinaryPredicate, typename _BinaryOperator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy,
+                                                             oneapi::dpl::__internal::__difference_t<_Range3>>
+sycl_reduce_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_keys,
+                       _Range4&& __out_values, _BinaryPredicate __binary_pred, _BinaryOperator __binary_op)
+{
+    using __diff_type = oneapi::dpl::__internal::__difference_t<_Range1>;
+    using __key_type = oneapi::dpl::__internal::__value_t<_Range1>;
+    using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+    using __flag_type = bool;
+
+    const auto __n = __keys.size();
+
+    constexpr int __vals_per_item = 2; // Each work item serially processes 2 items. Best observed performance on gpu
+
+    ::std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+
+    // adjust __wgroup_size according to local memory limit
+    __wgroup_size = oneapi::dpl::__internal::__max_local_allocation_size(
+        ::std::forward<_ExecutionPolicy>(__exec), sizeof(__key_type) + sizeof(__val_type), __wgroup_size);
+
+    int __n_groups = 1 + std::ceil(__n / (__wgroup_size * __vals_per_item));
```

Comment:
I think that required to check all evaluations of __n_groups and __n_groups type: my assumption that int may be incorrect type for these cases.

Reply:
Changed __n_groups to ::std::size_t,  Thanks.